### PR TITLE
[SPARK-47451][SQL] Support to_json(variant).

### PR DESCRIPTION
**Pull request by chenhao-db**
### What changes were proposed in this pull request?

This PR adds the functionality to format a variant value as a JSON string. It is exposed in the `to_json` expression by allowing the variant type (or a nested type containing the variant type) as its input.

### How was this patch tested?

Unit tests that validate the `to_json` result. The input includes both `parse_json` results and manually constructed bytes.
Negative cases with malformed inputs are also covered.

Some tests disabled in https://github.com/apache/spark/pull/45479 are re-enabled.

### CODE REVIEW

File: common/variant/src/main/java/org/apache/spark/variant/MalformedVariantException.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+package org.apache.spark.variant;
+
+import scala.collection.immutable.Map$;
+
+import org.apache.spark.QueryContext;
+import org.apache.spark.SparkRuntimeException;
+
+public class MalformedVariantException extends SparkRuntimeException {
+  public MalformedVariantException() {
+    super("MALFORMED_VARIANT", Map$.MODULE$.<String, String>empty(), null, new QueryContext[]{}, "");
```
Comment (gene-db) [inline common/variant/src/main/java/org/apache/spark/variant/MalformedVariantException.java:27]: I think the linter is complaining about this long line.

```
Error:  src/main/java/org/apache/spark/variant/MalformedVariantException.java:[27] (sizes) LineLength: Line is longer than 100 characters (found 101).
```

**Review by gene-db — CHANGES_REQUESTED** [review]

**Comment by gene-db** [comment]
@HyukjinKwon @cloud-fan Could you take a look at this? Thanks!

File: common/variant/src/main/java/org/apache/spark/variant/MalformedVariantException.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package org.apache.spark.variant;
+
+import scala.collection.immutable.Map$;
+
+import org.apache.spark.QueryContext;
+import org.apache.spark.SparkRuntimeException;
+
+public class MalformedVariantException extends SparkRuntimeException {
```
Comment (cloud-fan) [inline common/variant/src/main/java/org/apache/spark/variant/MalformedVariantException.java:25]: Unless we need to catch this certain exception to do something, it's recommended to simply use `SparkException` or `SparkRuntimeException` (if unchecked exception is needed) with error classes.

File: common/variant/src/main/java/org/apache/spark/variant/VariantConstructorSizeLimitException.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import org.apache.spark.QueryContext;
+import org.apache.spark.SparkRuntimeException;
+
+/**
+ * An exception indicating that an external caller tried to call the Variant constructor with value
+ * or metadata exceeding the 16MiB size limit. We will never construct a Variant this large, so it
+ * should only be possible to encounter this exception when reading a Variant produced by another
+ * tool.
+ */
+public class VariantConstructorSizeLimitException extends SparkRuntimeException {
```
Comment (cloud-fan) [inline common/variant/src/main/java/org/apache/spark/variant/VariantConstructorSizeLimitException.java:31]: ditto

File: common/variant/src/main/java/org/apache/spark/variant/MalformedVariantException.java (hunk 2 above)
Comment (cloud-fan) [reply common/variant/src/main/java/org/apache/spark/variant/MalformedVariantException.java:25]: end users should distinguish errors by error class, not different java exception class.

**Review by cloud-fan — APPROVED** [review]

File: common/variant/src/main/java/org/apache/spark/variant/MalformedVariantException.java (hunk 2 above)
Comment (chenhao-db) [reply common/variant/src/main/java/org/apache/spark/variant/MalformedVariantException.java:25]: Makes sense. I don't think we need to catch this certain exception.

**Review by gene-db — APPROVED** [review]

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Please hold on this PR because we need to resolve the following.

- https://github.com/apache/spark/pull/45479#pullrequestreview-1946939461

**Comment by dongjoon-hyun** [comment]
I made a PR. This PR is blocked by the following.
- #45685

**Comment by chenhao-db** [comment]
I have also performed renaming in this PR. It should be good to merge now.

File: common/variant/src/main/java/org/apache/spark/types/variant/VariantUtil.java [hunk 4]
```diff
@@ -141,4 +148,295 @@ public static byte arrayHeader(boolean largeSize, int offsetSize) {
...
+          case LONG_STR:
+            return Type.STRING;
+          default:
+            throw malformedVariant();
+        }
+    }
+  }
+
+  static IllegalStateException unexpectedType(Type type) {
+    return new IllegalStateException("Expect type to be " + type);
```
Comment (cloud-fan) [inline common/variant/src/main/java/org/apache/spark/types/variant/VariantUtil.java:259]: if this should not happen, we should throw `SparkException.internalError`

**Review by cloud-fan — APPROVED** [review]

File: common/variant/src/main/java/org/apache/spark/types/variant/VariantUtil.java (hunk 4 above)
Comment (cloud-fan) [reply common/variant/src/main/java/org/apache/spark/types/variant/VariantUtil.java:259]: We can address this minor issue in the next VARIANT PR.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**